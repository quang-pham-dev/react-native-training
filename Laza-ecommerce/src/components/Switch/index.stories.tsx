import React from 'react';
import { storiesOf } from '@storybook/react-native';
import CenterView from '../../../storybook/stories/CenterView';
import Switch from './index';

storiesOf('Switch', module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add('Default', () => <Switch />)
  .add('With False Value', () => <Switch value={false} />);
