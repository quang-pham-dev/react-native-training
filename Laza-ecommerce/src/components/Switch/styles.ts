import { Platform, StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  switch: {
    transform:
      Platform.OS === 'android'
        ? [{ scaleX: 0.9 }, { scaleY: 0.9 }]
        : [{ scaleX: 0.6 }, { scaleY: 0.6 }]
  }
});

export default styles;
