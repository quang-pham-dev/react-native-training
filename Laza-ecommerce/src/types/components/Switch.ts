export interface ISwitchProps {
  type?: 'android' | 'ios';

  value?: boolean;

  disable?: boolean;

  onToggle?: (newValue: boolean) => void;
}
