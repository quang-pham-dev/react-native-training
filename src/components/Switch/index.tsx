import React, {useCallback} from 'react'

// Libs
import {Switch as SwitchLIB} from 'react-native-switch'

// Themes
import {Colors} from '@themes'

type SwitchProps = {
  value?: boolean
  disabled?: boolean
}

const Switch = ({value = true, disabled = false}: SwitchProps) => {
  const [isEnabled, setIsEnabled] = React.useState(value)

  const handleToggle = useCallback((event: boolean) => {
    setIsEnabled(event)
  }, [])

  return (
    <SwitchLIB
      value={isEnabled}
      onValueChange={handleToggle}
      circleSize={16}
      barHeight={18}
      disabled={disabled}
      circleBorderWidth={0}
      backgroundActive={Colors.palette.emerald}
      backgroundInactive={Colors.palette.alto}
      circleActiveColor={Colors.palette.white}
      circleInActiveColor={Colors.palette.white}
      changeValueImmediately={true}
      renderActiveText={false}
      renderInActiveText={false}
      switchLeftPx={2}
      switchRightPx={2}
      switchWidthMultiplier={2}
      switchBorderRadius={30}
    />
  )
}

export default Switch
