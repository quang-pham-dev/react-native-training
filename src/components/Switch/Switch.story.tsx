import * as React from 'react'
import {storiesOf} from '@storybook/react-native'

import {StoryScreen, Story, UseCase} from '../../../storybook/views'

import Switch from './index'

declare let module: NodeModule

storiesOf('Switch', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Default', () => (
    <Story>
      <UseCase text="Default">
        <Switch />
      </UseCase>
    </Story>
  ))
  .add('Disabled', () => (
    <Story>
      <UseCase text="Disabled">
        <Switch disabled value={false} />
      </UseCase>
    </Story>
  ))
