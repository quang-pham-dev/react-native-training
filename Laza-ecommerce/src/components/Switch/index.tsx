import React, { useCallback } from 'react';
import { View, Switch as RNSwitch } from 'react-native';

// Themes
import Colors from 'themes/Colors';

// Types
import { ISwitchProps } from 'types/components/Switch';

// Styles
import styles from './styles';

const Switch = ({ onToggle, value = true }: ISwitchProps) => {
  const [isEnabled, setIsEnabled] = React.useState(value);

  const handleToggle = useCallback((event: boolean) => {
    setIsEnabled(event);
  }, []);

  const trackColor = {
    false: Colors.lightGray,
    true: Colors.lightGreen
  };

  return (
    <View>
      <RNSwitch
        style={styles.switch}
        trackColor={trackColor}
        thumbColor={isEnabled ? Colors.white : Colors.lightGray}
        ios_backgroundColor={Colors.lightGray}
        onValueChange={handleToggle}
        value={isEnabled}
      />
    </View>
  );
};

export default Switch;
