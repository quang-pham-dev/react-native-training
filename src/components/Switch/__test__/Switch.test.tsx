import React from 'react'

// LIBS
import {render} from '@testing-library/react-native'

// Components
import Switch from '@components/Switch'

describe('Switch Component', () => {
  const props = {
    value: true,
    disabled: false,
  }

  const tree = render(<Switch {...props} />)
  test('Should render correctly', () => {
    const component = tree.toJSON()
    expect(component).toMatchSnapshot()
  })
  test('Should render correctly with custom props', () => {
    const newProps = {
      ...props,
      disabled: true,
    }

    const component = render(<Switch {...newProps} />).toJSON()
    expect(component).toMatchSnapshot()
  })
})
