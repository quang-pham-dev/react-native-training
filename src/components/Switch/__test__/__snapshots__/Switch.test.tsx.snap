// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Component Should render correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(52, 199, 89, 1)",
      "borderRadius": 30,
      "height": 18,
      "width": 32,
    }
  }
  testID={null}
>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "left": 8,
        "width": 32,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderColor": "rgba(100, 100, 100, 1)",
          "borderRadius": 8,
          "borderWidth": 0,
          "height": 16,
          "justifyContent": "center",
          "width": 16,
        }
      }
    />
  </View>
</View>
`;

exports[`Switch Component Should render correctly with custom props 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(52, 199, 89, 1)",
      "borderRadius": 30,
      "height": 18,
      "width": 32,
    }
  }
  testID={null}
>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "left": 8,
        "width": 32,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderColor": "rgba(100, 100, 100, 1)",
          "borderRadius": 8,
          "borderWidth": 0,
          "height": 16,
          "justifyContent": "center",
          "width": 16,
        }
      }
    />
  </View>
</View>
`;
